src/librustc_typeck/diagnostics.rs

```diff
@@ -108,6 +108,64 @@ LinkedList::new() += 1;
 ```
 "##,
 
+E0072: r##"
+When defining a recursive struct or enum, any use of the type being defined
+from inside the definition must occur behind a pointer (like `Box` or `&`).
+This is because structs and enums must have a well-defined size, and without
+the pointer the size of the type would need to be unbounded.
+
+Consider the following erroneous definition of a type for a list of bytes:
+
+```
+// error, illegal recursive struct type
+struct ListNode {
+    head: u8,
+    tail: Option<ListNode>,
+}
+```
+
+This type cannot have a well-defined size, because it needs to be arbitrarily
+large (since we would be able to nest `ListNode`s to any depth). Specifically,
+
+```
+size of ListNode = 1 byte for head
+                 + 1 byte for the discriminant of the Option
+                 + size of ListNode
+```
+
+One way to fix this is by wrapping `ByteList` in a `Box`, like so:
+
+```
+struct ByteList {
+    head: u8,
+    tail: Option<Box<ByteList>>,
+}
+```
+
+This works because `Box` is a pointer, so its size is well-known.
+"##,
+
+E0073: r##"
+You cannot define a struct (or enum) `Foo` that requires an instance of `Foo`
+in order to make a new `Foo` value (think about it: if making an instance of
+`Foo` requires an already-existing instance of `Foo`, how does the first
+instance get made?).
```

Comment:
Perhaps this parenthetical could be made just its own sentence instead? Perhaps:

> You cannot define a struct (or enum) `Foo` that requires an instance of `Foo` in order to make a new `Foo` value. For example there is no way the first instance of `Foo` could be made to initialize another instance!